openedx/core/lib/xblock_keys/bundle_def.py

```diff
@@ -0,0 +1,139 @@
+import base64
+from uuid import UUID
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import DefinitionKey
+
+
+class BundleDefinitionLocator(DefinitionKey):
+    """
+    Implementation of the DefinitionKey type, for XBlock content stored in
+    Blockstore bundles (.olx files).
+
+    Each BundleDefinitionLocator holds the following data
+        bundle UUID (e.g. fc4d4d8d-60b3-4aad-b9f1-4ea738505d13)
+        OLX file path (e.g. /unit-welcome/welcome.olx)
+        Block type (of some block within the OLX file, e.g. 'html')
+        Block definition ID (of some block within the OLX file, e.g. 'intro')
+    
+    Note that we do not store bundle slugs (rather UUIDs), because the slugs
+    may change. These definition keys are more for internal use than for
+    displaying to users.
+    """
+    CANONICAL_NAMESPACE = 'olx-v1'
+    KEY_FIELDS = ('bundle_uuid', 'olx_path', 'block_type', 'definition_id')
+    __slots__ = KEY_FIELDS
+    CHECKED_INIT = False
+
+    def __init__(self, bundle_uuid, olx_path, block_type, definition_id):
+        if not isinstance(bundle_uuid, UUID):
+            bundle_uuid = UUID(bundle_uuid)
+        if len(olx_path) < 2 or olx_path[0] != '/' or olx_path.find(':') != -1:
+            raise ValueError("Invalid OLX path: {}".format(olx_path))
+        super(BundleDefinitionLocator, self).__init__(
+            bundle_uuid=bundle_uuid,
+            olx_path=olx_path,
+            block_type=block_type,
+            definition_id=definition_id,
+        )
+
+    @classmethod
+    def _olx_path_shortened(cls, olx_path):
+        """
+        Convert an arbitrary OLX path to a potentially shorter format.
+        The result is shorter if the olx path is in the standard format
+        /block_type-block_id/block_id.olx
+
+        The result is a string like "block_type/block_id" or the original string.
+
+        Returns a tuple: short_string, type_part or None, id_part or None
+        """
+        if len(olx_path) < 2 or olx_path[0] != '/':
+            raise ValueError("An OLX path must start with / and contain a filename")
+        if olx_path.endswith('.olx'):
+            id_part = olx_path[:-4].rpartition('/')[2]  # from "/foo/bar.olx", get "bar"
+            type_part = olx_path[1:].partition('-' + id_part + '/')[0] # from "/foo-bar/bar.olx", get "foo"
+            short_path = type_part + '/' + id_part
+            if cls._olx_path_expanded(short_path)[0] == olx_path:
+                return (short_path, type_part, id_part)
+        return (olx_path, None, None)
+
+    @classmethod
+    def _olx_path_expanded(cls, olx_path_shortened):
+        """
+        Convert a potentially shortened OLX path to the full expanded form
+        The result is a string like "/block_type-block_id/block_id.olx"
+
+        Returns a tuple: olx_path, type_part or None, id_part or None
+        """
```

Comment:
I appreciate that you've made the default format quite compact given all the data that's encoded inside. But is it necessary to encode the file location in the key at all? We could have some in-proc plugin on the Blockstore side that can examine new versions and build lookups/metadata that will take a `(bundle_uuid, block_type, locally_unique_identifier)` and return you the appropriate file. If that's necessary at all -- I imagine most of the time we read the OLX for a specific Bundle, we'd be greedily reading all of the OLX for that Bundle.

I ask mostly because I could see things moving around pretty easily -- files getting renamed to more appropriate things, items being moved between units, etc. And I'd be concerned if we were storing any user state against that kind of definition reference.

Reply:
Yeah, there's a lot to consider there and I'm totally open to changing this around completely.

I guess I'm really operating with the assumption that the operation "find the XBlock with `url_name` `X` which may be inside any OLX file in the given bundle" is very expensive to do, so I'm trying to avoid that. (Though maybe that will be rare in practice?) Yes, we could add a plugin that maintains a cache of all that, but that's yet another layer of state that needs to be maintained and yet another plugin we need to write, so I think it could be cool if we can avoid that by encoding more data in the runtime's keys etc.

> I ask mostly because I could see things moving around pretty easily -- files getting renamed to more appropriate things, items being moved between units, etc. And I'd be concerned if we were storing any user state against that kind of definition reference.

Hmm, OK, that's a good point. But if we try to allow file renames while still considering the "true identity" of the block to be preserved, there are some challenges:
* Other bundles which Link to the .olx file will now be broken. (I'm assuming that bundles which want to refer to other OLX files will do so by path e.g. `<vertical src="bundle:/_links/b2/other.olx"/>`, which seems natural for blockstore and not by definitionkey or {bundle_uuid, url_key} references?)
* We have to do a lot more work to ensure `url_name`s are unique. If we consider the file path part of the definition key, we only need to check `url_name` uniqueness within individual OLX files upon import (cheap, easy) vs. scanning the entire bundle.

Reply:
@ormsbee Yeah, after thinking about this more, I do think what you're saying makes sense.

So is it reasonable to update these definition keys to store `bundle_uuid`, `block_type`, and `url_name` only? And consider those three pieces of data both necessary and sufficient to identify any XBlock definition in Blockstore?

Reply:
And re DAGs, I'd like to avoid them but I think that we can allow _definition_ reuse in a course, so that the same definition appears in multiple places, but each place is considered a distinct usage.

Reply:
@bradenmacdonald: Yeah, I think `(bundle_uuid, block_type, url_name)` is sufficient.

Reply:
1. Will we not need the bundle `version_num` as well to go from key to a definition file?
2. Is it okay to have XBlocks of different types with the same `url_name` in a bundle? The `block_type` in the set implies that.

Reply:
> Will we not need the bundle version_num as well to go from key to a definition file?

I actually think we won't need it. The studio runtime can always load from the draft version. The LMS runtime won't load directly from blockstore anyways, and versioning is handled by Blockstore, when content from a specific version of another bundle is Linked into the course bundle.

> Is it okay to have XBlocks of different types with the same url_name in a bundle? The block_type in the set implies that.

Yeah, AFAIK it's always been the case in Open edX that only the pair of block_type+url_name is unique, not the url_name alone. I'd love to make url_name unique on its own, but that may add difficulties when importing existing content.

Reply:
> Yeah, AFAIK it's always been the case in Open edX that only the pair of block_type+url_name is unique, not the url_name alone. I'd love to make url_name unique on its own, but that may add difficulties when importing existing content.

Ah okay. It depends on whether it brings any advantages? The XBlocks that are going to get moved to Bundles are not going to retain their keys so any references from XBlocks in other Bundles will have to be updated anyway. So updating the `url_name` doesn't seem too much of a problem. If there is a clash one solution would be to just prepend the block type to it.

Reply:
Given that the vast majority of `url_name` entries are generated from Mongo object IDs, I'm guessing this would probably work out. We'd have to watch out for are the import/export workflow courses though. 😢 

If the eventual outcome is to replace Modulestore for the Studio use case and migrate all course content, then we have to have a way of preserving usage keys.